ons-spark/spark-app/spark_application_and_ui.md

```diff
@@ -0,0 +1,429 @@
+
+As we're running in local mode with 2 cores (or threads), the processing of these 200 tasks took place on the driver. The tasks seem to show relatively poor performance, which is indicated by the non-green colours in the timeline. Also by comparing the *Duration* (useful processing time) metrics with that of *Scheduler Delay* and *Task Deserialization Time* in the *Summary Metrics* table. Would we get better performance if the `aggregated_spark` DataFrame had fewer partitions? Let's try to improve the performace by reducing the number of partitions from 200 to 2.
+````{tabs}
+```{code-tab} py
+aggregated_spark = rescue.groupBy("cost_group").agg(F.count("incident_number").alias("count")).coalesce(2)
+aggregated_pandas = aggregated_spark.toPandas()
+```
+
+```{code-tab} r R
+
+aggregated_spark <- rescue %>% 
+    dplyr::group_by(cost_group) %>%
+    dplyr::summarise(count=n()) %>%
+    sparklyr::sdf_coalesce(2)
+
+                                    
+aggregated_r <- aggregated_spark %>% 
+    sparklyr::collect()
+
+```
+````
+
+```plaintext
+```
+Now navigate to the details of the latter stage of the latest job within the UI and you will see something similar to the screenshot below. Of course, the all important metric to compare is the time taken to complete the stage. Previously, using 200 partitions the stage time was 0.6 seconds, using 2 partitions the stage time was 0.3 seconds. Looking at this page we can see why.
+
+![Task timeline for fewer tasks showing better performance](../images/spark_app_2_tasks.png)
+
+Note that a 0.3s improvement is nothing to brag about, but a x2 processing speed isn't bad.
+
+Again, the processing took place on the driver with 2 cores, but this time there was one task per core. There's much more green visible in this timeline and comparing *Scheduler Delay* and *Task Deserialization Time* with *Duration* in the *Summary Metrics* tells a very different story.
+
+The important point is that we are processing a small amount of data and therefore should reduce the partitioning. With 200 partitions a lot of time was spent scheduling tasks and (de)serializing data. By putting our small DataFrame into fewer partitions we spent more time on useful processing and decareased the overall processing time.
+
+## Summary
+
+**Spark application hierarchy**
+
+    - Application 
+        - Job 
+            - Stage 
+                - Task
+
+
+**Use the Spark UI to**
+- debug slow or crashing applications
+- investigate what Spark is doing "under the hood"
+
+
+**Tips**
+- Look out for task colours, green is generally good
+- Use online documentation for more information about the UI
+- Lots of small partitions is an inefficient strategy to process data
+- Matching up the executed code with the job number in the UI is difficult, the description starts with the action used to initiate that job. You can also customise the job description to track jobs easier. See **Job Decription tip** for more information.
+
+## Exercises
```

Comment:
I think we should remove the exercises from the individual notebooks and put them in another article. What do you think?